libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +244,158 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
+						<userDefinedConstructor>(<params>)
+					</hasUserDefinedConstructor>
+				}
+			)");
+			string params;
+			if (contract->constructor())
+				for (ASTPointer<VariableDeclaration> const& varDecl: contract->constructor()->parameters())
+					params += (params.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl).commaSeparatedList();
+			t("params", params);
+			string baseParamsString = listAllParams(baseConstructorParams);
+			t("baseParams", baseParamsString);
+			t("comma", !params.empty() && !baseParamsString.empty() ? ", " : "");
+			t("functionName", implicitConstructorName(*contract));
+			pair<string, map<ContractDefinition const*, string>> evaluatedArgs = evaluateConstructorArguments(*contract);
+			baseConstructorParams.insert(evaluatedArgs.second.begin(), evaluatedArgs.second.end());
+			t("initBaseArguments", evaluatedArgs.first);
+			t("hasNextConstructor", i < _contract.annotation().linearizedBaseContracts.size() - 1);
+			if (i < _contract.annotation().linearizedBaseContracts.size() - 1)
+			{
+				ContractDefinition const* nextContract = _contract.annotation().linearizedBaseContracts[i + 1];
+				t("nextConstructor", implicitConstructorName(*nextContract));
+				t("nextParams", listAllParams(baseConstructorParams, nextContract));
+			}
+			t("initStateVariables", initStateVariables(*contract));
+			FunctionDefinition const* constructor = contract->constructor();
+			t("hasUserDefinedConstructor", constructor);
+			if (constructor)
+				t("userDefinedConstructor", m_context.enqueueFunctionForCodeGeneration(*constructor));
 
-			<assignToParams> <abiDecode>(memoryDataOffset, add(memoryDataOffset, argSize))
+			return t.render();
+		});
+	}
+}
 
-			<constructorName>(<params>)
-		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.enqueueFunctionForCodeGeneration(*constructor));
+string IRGenerator::constructorCode(ContractDefinition const& _contract)
+{
+	solAssert(!_contract.isLibrary(), "Tried to initialize state variables of library.");
 
-		out << t.render();
+	Whiskers creationCode(R"(
+		<callValueCheck>
+		<?hasParams>
+			let <params> := <copyArgumentsToMemory>()
+		</hasParams>
+		<constructor>(<params>)
+	)");
+	FunctionDefinition const* constructor = _contract.constructor();
+	creationCode("callValueCheck", constructor && !constructor->isPayable() ? callValueCheck() : "");
+	creationCode("hasParams", constructor && !constructor->parameters().empty());
+	string params;
+	if (constructor && !constructor->parameters().empty())
+	{
+		params = suffixedVariableNameList(
+			"param_" + m_context.functionName(*_contract.constructor()) + "_",
```

Comment:
```suggestion
			"param_",
```
- it is a local variable, so it does not have to be that specific.